File: 2023/ruby/p08.rb
input = IO.foreach('../data/08.txt').map(&:strip)

ops = input[0]

left = {}
right = {}
input[2..-1].each{|line| 
  parts = line.split(/[\s,\(\)]/)
  left[parts[0]] = parts[3]
  right[parts[0]] = parts[5]
}

def find(start, target, left, right, ops)
  node = start
  steps = 0
  
  until node.end_with? target
    d = ops[steps % ops.size]
    steps += 1
    node = d == 'L' ? left[node] : right[node]
  end
  steps
end

# part 1
p find('AAA', 'ZZZ', left, right, ops)

# part 2
starts = left.keys.filter{|node| node.end_with?'A'}
p starts.map{|start_node| find(start_node, 'Z', left, right, ops)}.reduce(:lcm)

File: 2023/nim/p08.nim
import strutils, sequtils, tables, math, re

let
    input = readFile("../data/08.txt").strip.splitLines
    ops = input[0]

var
    left = initTable[string, string]()
    right = initTable[string, string]()
for line in input[2..^1]:
    let parts = line.split(re"[\s,\(\)]")
    left[parts[0]] = parts[3]
    right[parts[0]] = parts[5]

proc findNode(start, target: string): int =
    var node = start
    var steps = 0
    while not node.endsWith(target):
        let d = ops[steps mod ops.len]
        steps += 1
        node = if(d == 'L'): left[node] else: right[node]
    steps

# part 1
echo findNode("AAA", "ZZZ")

# part 2
let starts = left.keys.toSeq.filterIt(it.endsWith("A"))
echo starts.mapIt(findNode(it, "Z").uint64).foldl(lcm(a, b))
